drivers/sensor/bosch/bmi08x/bmi08x_accel.c

```diff
@@ -454,16 +454,33 @@ static int bmi08x_temp_channel_get(const struct device *dev, struct sensor_value
 {
 	uint16_t temp_raw = 0U;
 	int32_t temp_micro = 0;
+	int16_t temp_int11 = 0;
 	int ret;
 
 	ret = bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_MSB, &temp_raw);
-	if (ret < 0) {
+	if (!ret) {
+		temp_int11 = (temp_raw & 0xFF) << 3;
+	} else {
 		return ret;
 	}
 
-	/* the scale is 1/2^5/LSB = 31250 micro degrees */
-	temp_micro = BMI08X_TEMP_OFFSET * 1000000ULL + temp_raw * 31250ULL;
+	if (temp_raw == 0x80) {
+		/* temperature invalid */
+		return -ENODATA;
+	}
 
+	ret = bmi08x_accel_word_read(dev, BMI08X_REG_TEMP_LSB, &temp_raw);
+	if (!ret) {
+		temp_int11 |= (temp_raw & 0xE0) >> 5;
+	} else {
+		return ret;
+	}
+	if (temp_int11 > 1023) {
+		temp_int11 -= 2048;
+	}
+	/* temp_int11 ranges in [-504, 496] */
+	/* the scale is 0.125°C/LSB = 125 micro degrees */
+	temp_micro = temp_int11 * 0.125 * 1000ULL + 23 * 1000000ULL;
```

Comment:
this is using double percision floating point, can do it without double percision?

Reply:
Should we eliminate the temp_micro intermediate variable and directly calculate val->val1 and val->val2 using these expressions:

/* temp_int11 ranges in [-504, 496] */
/* resolution: 125 micro degrees per count */
val->val1 = 23 + ((temp_int11 * 125) / 1000000);
val->val2 = (temp_int11 * 125) % 1000000;

Or just:
temp_micro = temp_int11 * 125 + 23 * 1000000;

Reply:
Okay. But at the same time I wonder doesn't the compiler optimize it automatically? thx!

Reply:
It wouldn't optimize something like that... but it looks good now thanks!
https://godbolt.org/z/E5WTvvxEK

Reply:
I see! thanks a lot! I used to think that GCC would automatically determine variable types and treat all constant numbers as the same type.